### PR TITLE
update headers

### PR DESCRIPTION
**Pull request by shuisman**
XHRs with 'Date' header are not allowed by W3.org > moving to x-usbl-date

Reviewers: @usabilla/feds @usabilla/plumbers 

### Changelog:
- Date header to x-usbl-date header

### Notes for Testing:
- _None_

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "usabilla-api",
-  "version": "1.0.0-beta.1",
+  "version": "1.0.0-beta.2",
```
Comment (spirosikmd) [inline package.json]: Version is better to be updated with another PR. First we get all the code that needs to go to another version in master and then we can create a PR to update version.

File: src/index.js [hunk 2]
```diff
@@ -262,30 +262,18 @@ class SignatureFactory {
...
-      this.url,
-      queryStr,
-      canonicalHeaders.date,
-      canonicalHeaders.host,
-      'date;host',
-      bodyHash
+      this.method || 'GET',               // HTTPRequestMethod
+      this.url,                           // CanonicalURI
+      this.queryParameters,               // CanonicalQueryString
+      `host:${this.host}`+ '\n' + `x-usbl-date:${this.dates.longdate}`+ '\n', // CanonicalHeaders
```
Comment (spirosikmd) [inline src/index.js]: These strings are already joined by `\n`. So you can change this to

```javascript
return [
  ...,
  this.queryParameters,
  `host:${this.host}`,
  `x-usbl-date:${this.dates.longdate}`,
  'host;x-usbl-date',
  ...
];
```

Makes it easier to read without the extra `\n`.

File: src/index.js [hunk 3]
```diff
@@ -262,30 +262,18 @@ class SignatureFactory {
...
+      'host;x-usbl-date',                 // SignedHeaders
```
Comment (spirosikmd) [inline src/index.js]: The `x-usbl-date` string is repeated many times in the `SignatureFactory` class. Maybe you can add it as an instance variable `this.dateHeader = 'x-usbl-date';` and then use it trough this class?

Comment (shuisman) [inline src/index.js]: hmm, this is difficult..
You suggestion does not hold, as the headers need to be in a specific place / order based on the name. Therefore if extracting the date header name,  one will loose the contextual connection to the order.
I thought about grouping all headers (and order) in one place, but this can be approached from different aspects, as some headers are signature related and others are not. Also the content of the header (the longdate) is defined in two places, which makes me sad.

I suggest to leave it as it is, and some extensive discussion with refactoring needs to be done to solve this.

File: package.json (hunk 1 above)
Comment (shuisman) [inline package.json]: agreed, will fix

File: src/index.js (hunk 2 above)
Comment (shuisman) [inline src/index.js]: My focus here is a line per property of the CanonicalString.
You suggestion breaks consistency as you define one property (the CanonicalHeaders) on two lines. Secondly, each property needs to end with a \n (done in the join) and for the headers in the CanonicalHeaders needs to end with a \n. Resulting in a double \n at the end of this property. Your suggestion breaks the functionality.

I suggest to leave it as this.

Comment (spirosikmd) [inline src/index.js]: Ok so you need the extra `\n`? How about this?

```javascript
return [
  ...,
  this.queryParameters,
  `host:${this.host}\n`,
  `x-usbl-date:${this.dates.longdate}\n`,
  'host;x-usbl-date',
  ...
];
```

This is much more readable. I don't really agree with the comments next to the lines as formatting is horrible. Would be more readable to put them at the top of this return.

File: src/index.js (hunk 3 above)
Comment (spirosikmd) [inline src/index.js]: I don't think this is as difficult as you think. It is basically extracting the string to a variable to avoid code duplication. Small steps toward refactoring!

File: src/index.js (hunk 2 above)
Comment (shuisman) [inline src/index.js]: Hm, this would also not work. From your perspective you want;
`return [
  ...,
  this.queryParameters,
  `host:${this.host}`,
  `x-usbl-date:${this.dates.longdate}\n`,
  'host;x-usbl-date',
  ...
];
`
This is weird, as the comma joins and adds a `\n` between the properties of the CanonicalString. Now you use that functionality to add a `\n` for a header inside a property of the CanonicalString. That break in consistency makes it more unreadable.

The explanation can be moved to the top, no problem. I copied this practise from usabilla dev PHP coding.

File: src/index.js (hunk 3 above)
Comment (shuisman) [inline src/index.js]: True, implementing your solution is not difficult.
But in structure, it would ease the way of problems as you disconnect the naming and the ordering.

Comment (shuisman) [inline src/index.js]: as discussed face-2-face > move to different PR

File: src/index.js (hunk 2 above)
Comment (shuisman) [inline src/index.js]: as discussed face-2-face > move out of array for readbility. New PR would resolve this.

File: src/index.js [hunk 4]
```diff
@@ -269,7 +269,7 @@ class SignatureFactory {
   canonicalString () {
 
     // CanonicalHeaders
-    CanonicalHeaders = [
+    let CanonicalHeaders = [
```
Comment (spirosikmd) [inline src/index.js]: Just don't use capital and not join

```javascript
let canonicalHeaders = `host:${this.host}\n` + `x-usbl-date:${this.dates.longdate}\n`
```
